File: fortran/src/main/java/de/irt/jfor/Mat.java
package de.irt.jfor;

/**
 * Created by IntelliJ IDEA.
 * User: stueken
 * Date: 20.10.17
 * Time: 11:54
 */
public class Mat {

    public final int nx;

    public final int ny;

    public Mat(int nx, int ny) {
        if(nx<1 || ny<1)
            throw new IllegalArgumentException("invalid  array index");

        this.nx = nx;
        this.ny = ny;
    }

    static int bound(int ix, int nx, String name) {
        if(ix<1 || ix>nx) {
            String error = String.format("index %s out of bounds %d: %d", name, ix, nx);
            throw new IndexOutOfBoundsException(error);
        }
        return ix-1;
    }

    protected int index(int ix, int iy) {
        return bound(ix, nx, "nx") + nx * bound(iy, ny, "ny");
    }
}

File: fortran/src/main/java/de/irt/kfor/Mat.kt
package de.irt.kfor

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  18.01.2019 10:34
 * modified by: $Author$
 * modified on: $Date$
 */
interface Mat {
    val ni : Int
    val nj : Int
    val len : Int get() = ni*nj

    fun index(i : Int, j : Int) = i-1 + ni*(j-1)
}